include/rcpputils/patch.hpp

```diff
@@ -0,0 +1,200 @@
+};
+
+template<size_t N, typename ReturnType,
+  typename ArgType0, typename ArgType1,
+  typename ArgType2, typename ArgType3,
+  typename ArgType4, typename ArgType5>
+struct patch_traits<N, ReturnType(ArgType0, ArgType1, ArgType2, ArgType3, ArgType4, ArgType5)>
+{
+  mmk_mock_define(
+    mock_type, ReturnType, ArgType0, ArgType1, ArgType2, ArgType3, ArgType4, ArgType5);
+};
+
+template<size_t N, typename SignatureT>
+struct trampoline;
+
+template<size_t N, typename ReturnT, typename ... ArgTs>
+struct trampoline<N, ReturnT(ArgTs...)>
+{
+  static ReturnT base(ArgTs... args)
+  {
+    return target(std::forward<ArgTs>(args)...);
+  }
+
+  static std::function<ReturnT(ArgTs...)> target;
+};
+
+template<size_t N, typename ReturnT, typename ... ArgTs>
+std::function<ReturnT(ArgTs...)>
+trampoline<N, ReturnT(ArgTs...)>::target;
+
+template<size_t N, typename SignatureT>
+auto prepare_trampoline(std::function<SignatureT> target)
+{
+  trampoline<N, SignatureT>::target = target;
+  return trampoline<N, SignatureT>::base;
+}
+
+template<size_t N, typename SignatureT>
+class patch;
+
+template<size_t N, typename ReturnT, typename ... ArgTs>
+class patch<N, ReturnT(ArgTs...)>
+{
+public:
+  using mock_type = typename patch_traits<N, ReturnT(ArgTs...)>::mock_type;
+
+  patch(const std::string & target, std::function<ReturnT(ArgTs...)> proxy)
+  : proxy_(proxy)
+  {
+    auto MMK_MANGLE(mock_type, create) =
+      patch_traits<N, ReturnT(ArgTs...)>::MMK_MANGLE(mock_type, create);
+    mock_ = mmk_mock(target.c_str(), mock_type);
+  }
+
+  patch(const patch &) = delete;
+  patch & operator=(const patch &) = delete;
+
+  patch(patch && other)
+  {
+    mock_ = other.mock_;
+    other.mock_ = nullptr;
+  }
+
+  patch & operator=(patch && other)
+  {
+    if (mock_) {
+      mmk_reset(mock_);
+    }
+    mock_ = other.mock_;
+    other.mock_ = nullptr;
+  }
+
+  ~patch()
+  {
+    if (mock_) {
+      mmk_reset(mock_);
+    }
+  }
+
+  patch & then_call(std::function<ReturnT(ArgTs...)> replacement) &
+  {
+    auto type_erased_trampoline =
+      reinterpret_cast<mmk_fn>(prepare_trampoline<N>(replacement));
+    mmk_when(proxy_(any<ArgTs>()...), .then_call = type_erased_trampoline);
+    return *this;
+  }
+
+  patch && then_call(std::function<ReturnT(ArgTs...)> replacement) &&
+  {
+    auto type_erased_trampoline =
+      reinterpret_cast<mmk_fn>(prepare_trampoline<N>(replacement));
+    mmk_when(proxy_(any<ArgTs>()...), .then_call = type_erased_trampoline);
+    return std::move(*this);
+  }
+
+private:
+  template<typename T>
+  T any() {return mmk_any(T);}
+
+  mock_type mock_;
+  std::function<ReturnT(ArgTs...)> proxy_;
+};
+
+template<size_t N, typename SignatureT>
+auto make_patch(const std::string & target, std::function<SignatureT> proxy)
+{
+  return patch<N, SignatureT>(target, proxy);
+}
+
+}  // namespace rcpputils
+
+#define RCPPUTILS_PATCH_IGNORE_OPERATOR(type, op) \
```

Comment:
I'm sure you'll add more info in comments, but what purpose does overloading these operators have? It's not clear from the macro name.

Reply:
For every argument type, a set of comparison operators must be defined (even if meaningless)  or otherwise `mimick`'s macro generated code won't compile. This macro attempts to be a handy way to generate these dummy overloads. Name is up for debate :sweat_smile: 

FWIW I tried to come up with a smarter mechanism to hide this but couldn't find a way.